grommet/grommet
# Prevent clicking inside a nested Drop from closing parent Drop

#### What does this PR do?

Intercepts and prevents the `mousedown` event from being propagated upwards from a `DropContainer`. The purpose behind this is to prevent interacting with a nested Drop from closing its parent, since the DOM elements are technically siblings, not in a child-parent relationship (this due to React portals being involved).

The fix, while simple, has a potentially unwanted side-effect of making it so that you cannot close a nested Drop by clicking inside its "parent" Drop. I think a solution which would solve that problem would be notably more complex and may require use of a context.

This caveat may be most evident if the user puts a Form inside a Drop, then adds a Select to that form. The user will not be able to close the select by clicking elsewhere in the Form.

I wanted to proceed with opening the PR though to get the conversation started on solutions to that problem. I'll keep thinking of the simplest way to mitigate the issue.

#### Where should the reviewer start?

I modified the Progressive Drop story in Storybook to use a nested drop to demonstrate the behavior. You should be able to interact with the button inside the final drop without closing all the Drops at once.

#### What testing has been done on this PR?

Manual testing via storybook.

#### How should this be manually tested?

1. Run "Progressive" Drop story
2. Click through the nested Drops
3. Verify that you can click / interact with the final drop without closing its parent

#### Any background context you want to provide?

#### What are the relevant issues?

#2741 

#### Screenshots (if appropriate)

#### Do the grommet docs need to be updated?

#### Should this PR be mentioned in the release notes?

Probably, as a bugfix.

#### Is this change backwards compatible or is it a breaking change?

The undesirable behavior around not being able to click inside the parent to close the child would be backwards incompatible.
